source/differentiation/ad/ad_helpers.cc

```diff
@@ -0,0 +1,783 @@
+            ExcMessage(
+              "Need to extract sensitivities in the order they're created."));
+        }
+
+      if (ADNumberTraits<ad_type>::is_taped == true)
+        {
+          Assert(active_tape() != types::invalid_tape_index,
+                 ExcMessage("Invalid tape index"));
+          Assert(is_recording() == true,
+                 ExcMessage("Only valid during recording"));
+        }
+
+      internal::Marking<ad_type>::independent_variable(
+        independent_variable_values[index],
+        index,
+        this->n_independent_variables(),
+        out);
+      touched_marked_independent_variables[index] = true;
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    void
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::
+      finalize_sensitive_independent_variables() const
+    {
+      // Double check that we've actually touched all DoFs
+      Assert(n_touched_independent_variables() == n_independent_variables(),
+             ExcMessage("Not all values of sensitivities have been recorded!"));
+
+      // This should happen only once
+      if (this->independent_variables.size() == 0)
+        {
+          this->independent_variables =
+            std::vector<ad_type>(this->n_independent_variables(),
+                                 dealii::internal::NumberType<ad_type>::value(
+                                   0.0));
+
+          // Indicate the sensitivity that each entry represents
+          for (unsigned int i = 0; i < this->n_independent_variables(); ++i)
+            this->mark_independent_variable(this->independent_variables[i], i);
+        }
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    void
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::
+      initialize_non_sensitive_independent_variable(
+        ad_type &          out,
+        const unsigned int index) const
+    {
+      if (ADNumberTraits<ad_type>::is_taped == true)
+        {
+          Assert(active_tape() != types::invalid_tape_index,
+                 ExcMessage("Invalid tape index"));
+        }
+      Assert(is_recording() == false,
+             ExcMessage("Only valid outside of recording"));
+
+      Assert(index < n_independent_variables(), ExcInternalError());
+      Assert(touched_independent_variables[index] == true, ExcInternalError());
+
+      out = independent_variable_values[index];
+    }
+
+
+
+    template <int                  dim,
+              enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType>
+    unsigned int
+    ADHelperBase<dim, ADNumberTypeCode, ScalarType>::
+      n_touched_independent_variables() const
+    {
+      return std::accumulate(touched_independent_variables.begin(),
+                             touched_independent_variables.end(),
+                             0u);
```

Comment:
this is hard to understand. it works because the data type of the array (`bool`) can be converted to `unsigned int`, but I think what you really want to do is `std::count (...begin(), ...end(), true)`